libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs

```diff
@@ -252,7 +253,8 @@ class EraTx era => EraUTxO era where
   -- for the transaction to be valid, once those scripts are evaluated
   getScriptsNeeded :: UTxO era -> TxBody era -> ScriptsNeeded era
 
-  -- | Extract the set of all script hashes that are needed for script validation.
+  witsVKeyNeeded :: UTxO era -> TxBody era -> GenDelegs (EraCrypto era) -> Set (KeyHash 'Witness (EraCrypto era))
```

Comment:
I used it to construct the set of witnesses for a transaction in the `mkTxWits` function. I also noticed that we had copied the whole `Utxow` rule just to swap out the `witsVKeyNeeded` function in Conway. I figured that it makes sense to have this as a class method anyways so we can better reuse code across eras.

Reply:
As we discussed in a call, this is far from being the only reason why the Utxow was recreated from scratch in almost every era. There are major differences in all of them.

Since this functionality is needed in tests it should be solved in tests. So, adding such function to a type class in a test suite is a totally sensible solution.

Also a request for the future. Could you please separate any serious changes like that, which affect the rules, into a separate smaller PR. Mixing restructures like that to the rules with refactoring of test suites could obfuscate potential bugs.